research/picai/preprocessing.py

```diff
@@ -0,0 +1,358 @@
+
+
+class PicaiCase(Case):
+    def __init__(
+        self,
+        scan_paths: Sequence[Path],
+        annotations_path: Path,
+        settings: PreprocessingSettings,
+    ) -> None:
+        """
+        Class representing a case from the PICAI dataset.
+
+        Args:
+            scan_paths (Sequence[Path]): The set of paths where the scans associated with the Case are located.
+            annotation_write_dir (Path): The path where the annotation associated with the Case is located.
+            settings (PreprocessingSettings): The settings determining how the case is preprocessed.
+        """
+        super().__init__(scan_paths, annotations_path, settings)
+
+    def read(self) -> None:
+        """
+        Reads in scans and annotation into the corresponding scans and annotation attribute.
+        """
+        assert len(self.scan_paths) != 0
+        self.scans = [sitk.ReadImage(path) for path in self.scan_paths]
+        self.annotation = sitk.ReadImage(self.annotations_path)
+
+    def write(self) -> Tuple[Sequence[Path], Path]:
+        """
+        Writes preprocessed scans and annotations from PICAI dataset to disk
+        and returns the scan file paths and annotation file path in a tuple.
+
+        Returns:
+            Tuple[Sequence[Path], Path]: A tuple in which the first entry is a sequence of file paths
+                for the scans and the second entry is the file path to the correpsonding annotation.
+        """
+        modality_suffix_map = {"t2w": "0000", "adc": "0001", "hbv": "0002"}
+        scan_paths = [path for path in sorted(self.scan_paths)]
+        preprocessed_scan_paths = []
+        for path, scan in zip(scan_paths, self.scans):
+            scan_filename = os.path.basename(path)
+            scan_filename_without_extension = scan_filename.split(".")[0]
+            suffix = modality_suffix_map[scan_filename_without_extension[-3:]]
+            preprocessed_scan_filename = scan_filename_without_extension[:-3] + suffix + ".nii.gz"
+            preprocessed_scan_path = Path(os.path.join(self.settings.scans_write_dir, preprocessed_scan_filename))
+            sitk.WriteImage(scan, preprocessed_scan_path, useCompression=True)
+            preprocessed_scan_paths.append(preprocessed_scan_path)
+
+        annotation_filename = os.path.basename(self.annotations_path)
+        preprocessed_annotation_path = Path(os.path.join(self.settings.annotation_write_dir, annotation_filename))
+        sitk.WriteImage(self.annotation, preprocessed_annotation_path, useCompression=True)
+
+        return preprocessed_scan_paths, preprocessed_annotation_path
+
+
+class PreprocessingException(Exception):
+    """
+    Custom exception to be thrown during a failed attempt at processing a Case.
+    """
+
+    pass
+
+
+class PreprocessingTransform(ABC):
+    """Abstract class that represents a transform to be applied to a case."""
+
+    @abstractmethod
+    def __call__(self, mri: Case) -> Case:
```

Comment:
Super minor, but perhaps we just call this argument case to be more generic than an MRI? If it is limited to an MRI though, we can leave it.

Reply:
Good idea, I think its not specific to MRI. Any 3D medical image so CT would also work. I will do some follow up work to further generalize these so they can represent more than just 3D medical images but started with that just cause its the case for the PICAI dataset